akka-actor/src/main/scala/akka/pattern/Patterns.scala

```diff
@@ -454,6 +455,76 @@ object Patterns {
+    scalaRetry(() => attempt.call, attempts)(context)
+  }
+
+  /**
+   * Returns an internally retrying [[java.util.concurrent.CompletionStage]]
+   * The first attempt will be made immediately, each subsequent attempt will be made immediately
+   * if the previous attempt failed.
+   * If attempts are exhausted the returned completion operator is simply the result of invoking attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent tries
+   * and therefore must be thread safe (not touch unsafe mutable state).
+   */
+  def retry[T](attempt: Callable[CompletionStage[T]], attempts: Int, ec: ExecutionContext): CompletionStage[T] = {
+    require(attempt != null, "Parameter attempt should not be null.")
+    scalaRetry(() => attempt.call().toScala, attempts)(ec).toJava
+  }
+
+  /**
+   * Returns an internally retrying [[scala.concurrent.Future]]
+   * The first attempt will be made immediately, each subsequent attempt will be made with a backoff time,
+   * if the previous attempt failed.
+   * If attempts are exhausted the returned future is simply the result of invoking attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent tries and
+   * therefore must be thread safe (not touch unsafe mutable state).
+   */
+  def retry[T](
+      attempt: Callable[Future[T]],
+      attempts: Int,
+      minBackoff: FiniteDuration,
+      maxBackoff: FiniteDuration,
+      randomFactor: Double,
+      scheduler: Scheduler,
+      context: ExecutionContext): Future[T] = {
+    require(attempt != null, "Parameter attempt should not be null.")
+    scalaRetry(() => attempt.call, attempts, minBackoff, maxBackoff, randomFactor)(context, scheduler)
+  }
+
+  /**
+   * Returns an internally retrying [[java.util.concurrent.CompletionStage]]
+   * The first attempt will be made immediately, each subsequent attempt will be made with a backoff time,
+   * if the previous attempt failed.
+   * If attempts are exhausted the returned future is simply the result of invoking attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent tries and
+   * therefore must be thread safe (not touch unsafe mutable state).
+   */
+  def retry[T](
+      attempt: Callable[CompletionStage[T]],
+      attempts: Int,
+      minBackoff: java.time.Duration,
+      maxBackoff: java.time.Duration,
+      randomFactor: Double,
+      scheduler: Scheduler,
+      ec: ExecutionContext): CompletionStage[T] = {
+    require(attempt != null, "Parameter attempt should not be null.")
+    require(minBackoff != null, "Parameter minBackoff should not be null.")
+    require(maxBackoff != null, "Parameter minBackoff should not be null.")
+    scalaRetry(() => attempt.call().toScala, attempts, minBackoff.asScala, maxBackoff.asScala, randomFactor)(
+      ec,
+      scheduler).toJava
+  }
```

Comment:
Same for these two, drop the Future one and add Java test coverage to make sure things works.

Reply:
WIll fix

Reply:
I have added the Java tests.
I think we should keep those two styles for now and drop or deprecate the `Future` ones only if we do a totally clean up in this file?

Reply:
I'd be ok to not drop or deprecate in this PR but I don't think adding methods that should be deprecated immediately is a good idea, so skip those (and keep existing)?